message.go

```diff
@@ -502,13 +502,25 @@ func (m *Message) Unmarshal(v interface{}) error {
 		}
 
 		dataField := dataStruct.Field(i)
-		if dataField.IsNil() {
-			dataField.Set(reflect.New(dataField.Type().Elem()))
-		}
+		switch dataField.Kind() { //nolint:exhaustive
+		case reflect.Chan, reflect.Func, reflect.Map, reflect.Pointer, reflect.UnsafePointer, reflect.Interface, reflect.Slice:
+			if dataField.IsNil() {
+				dataField.Set(reflect.New(dataField.Type().Elem()))
+			}
 
-		err = messageField.Unmarshal(dataField.Interface())
-		if err != nil {
-			return fmt.Errorf("failed to get value from field %d: %w", fieldIndex, err)
+			err = messageField.Unmarshal(dataField.Interface())
+			if err != nil {
+				return fmt.Errorf("failed to get value from field %d: %w", fieldIndex, err)
+			}
+		default: // Native types
+			// using new reflect value that has pointer instead of input data field
+			dummy := reflect.New(dataField.Type()).Elem()
+			err = messageField.Unmarshal(dummy)
+			if err != nil {
+				return fmt.Errorf("failed to get value from field %d: %w", fieldIndex, err)
+			}
+
+			dataField.Set(dummy)
```

Comment:
If you remove `dummy` and keep code like this, all tests will work:

```suggestion
			err = messageField.Unmarshal(dataField)
			if err != nil {
				return fmt.Errorf("failed to get value from field %d: %w", fieldIndex, err)
			}
```

it's because in `dataField := dataStruct.Field(i)` `dataField` is `reflect.Value`, which we are using inside each field.